### PR TITLE
List clusters stale results fix

### PR DESCRIPTION
**Pull request by akarukappadath**
Background: When calling the list clusters endpoint, it would fail to include recently created clusters because of a call to Sam during cluster creation that is not transactional. This fix returns clusters that were created by the user as well as any clusters returned by Sam as those that the user should be able to see. 

Testing: This was manually tested using the Terra UI. The problem presented itself in a variety of ways, including https://github.com/databiosphere/leonardo/issues/821 and https://github.com/DataBiosphere/leonardo/issues/844. To manually test this, I reproduced these situations and made sure that it had resolved. 

Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/leonardo/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get a thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/AuthProviderSpec.scala [hunk 1]
```diff
@@ -119,10 +119,6 @@ class AuthProviderSpec extends FreeSpec with ScalatestRouteTest with Matchers wi
       clusterGetResponseException shouldBe a [ClusterNotFoundException]
       clusterGetResponseException.asInstanceOf[ClusterNotFoundException].statusCode shouldBe StatusCodes.NotFound
 
-      //list all clusters should be fine, but empty
```
Comment (akarukappadath) [inline src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/AuthProviderSpec.scala]: This section of the test will not pass, as all clusters created by a user will be returned in a list clusters call.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala [hunk 2]
```diff
@@ -527,8 +527,12 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig,
     for {
       paramMap <- processListClustersParameters(params)
       clusterList <- dbRef.inTransaction { da => da.clusterQuery.listByLabels(paramMap._1, paramMap._2, googleProjectOpt) }
-      visibleClusters <- authProvider.filterUserVisibleClusters(userInfo, clusterList.map(c => (c.googleProject, c.clusterName)).toList)
+      samVisibleClusters <- authProvider.filterUserVisibleClusters(userInfo, clusterList.map(c => (c.googleProject, c.clusterName)).toList)
+      // Making the assumption that users will always be able to access clusters that they create
```
Comment (MatthewBemis) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: i think this is a safe assumption to make. per [sam's config](https://github.com/broadinstitute/sam/blob/develop/src/main/resources/reference.conf#L286-L317), we don't even give people the permission to add/remove others from a cluster so once somebody creates a cluster they should never be removed.

at the very least this isn't driving access so even if the above weren't true, it wouldn't let anybody access data/clusters they shouldn't be able to see or didn't know about already.

**Review by MatthewBemis — APPROVED** [review]
:+1:

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala [hunk 3]
```diff
@@ -527,8 +527,12 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig,
...
+      // Fix for https://github.com/DataBiosphere/leonardo/issues/821
+      createdClusters <- dbRef.inTransaction { da => da.clusterQuery.listClustersByCreator(userInfo.userEmail).map(c => c.toList.map(cluster => (cluster.googleProject, cluster.clusterName))) }
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: Do we need to make an additional query here? We already have the full list of clusters in the project in the `clusterList` variable. Could you just do:
```
val visibleClusters = samVisibleClusters ::: clusterList.filter(_.creator == userInfo.userEmail)
```

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala [hunk 4]
```diff
@@ -164,6 +164,12 @@ trait ClusterComponent extends LeoComponent {
       }
     }
 
+    def listClustersByCreator(creator: WorkbenchEmail): DBIO[Seq[Cluster]] = {
+      clusterQuery.filter { _.creator === creator.value }.result map { recs =>
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala]: If adding this query we should also add an index on the `creator` column -- but as noted below I'm not sure it's needed.

**Review by rtitle — CHANGES_REQUESTED** [review]
See 1 comment above about the additional DB query

**Comment by akarukappadath** [comment]
jenkins retest

File: src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/AuthProviderSpec.scala (hunk 1 above)
Comment (rtitle) [reply src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/AuthProviderSpec.scala]: Is there a way to actually test the new behavior?

**Comment by rtitle** [comment]
The latest commit looks good -- I would like to see a new unit test that validates the new behavior (e.g. listClusters returns user-owned clusters even if not returned by the auth provider).

**Review by rtitle — APPROVED** [review]
💯

**Pull request merged**